### PR TITLE
Added Variable Unit Quality to Unit Markets and Updated Unit Market Logic

### LINKED ISSUE #2085: RFE: AtB: Vary unit quality in the Unit Market
It'd be cool to have some variety in the quality of the units offered for sale in the Unit Market (along with prices to match). So that  you have choice between a pristine mech from the factory that costs more, or a discounted option from the Mercenary Market but the parts are mostly A/B rating.

This could be extended furthered by have some units be randomly damaged (e.g. missing engine, gyro, weapons, limbs, etc.) for an even greater discount. (Could even be hidden: Surprise, that Black Market Marauder doesn't have an engine!)

### PR DESCRIPTION
**Pull request by IllianiCBT**
Requires: #4150 and #4158

This PR enhances the functionality of the AtB Unit Market with the following objectives:

1. Streamline the process for adding new units for sale. Upon reviewing the code, I noticed numerous dice rolls that lacked clarity regarding their comparison to other market types. I standardized these rolls, introducing human-readable rarity levels to provide maintainers with clear insights into the rarity of units within any given market.

2. Introduce greater depth to the markets by considering campaign faction and contract state more effectively. Previously, many markets defaulted to the employer faction, potentially leading to difficulties for House Units in maintaining their faction's units between contracts, or problems when no employer is available.

3. Associate unit quality with market type. While initially, I incorrectly believed this had always been the intent, it seemed too good an idea to abandon. This is applicable for campaigns with and without random unit quality enabled. In the event random unit quality is enabled, the market types will apply different modifiers to the quality roll. In the event this option is disabled, the unit quality will be fixed based on market type.

Additionally, I have provided comprehensive documentation for unit markets and renamed the 'AtB Stuff' folder to 'Against the Bot'. You can access the documentation [here](https://docs.google.com/document/d/1yBzdo2cjESk7sPrfzC5kE59XJ9pd9XDV1P_nWLUlLPE/edit?usp=sharing).

## Update
This now includes a bug fix for the Unit Market that otherwise prevents Artillery from spawning in the unit market. This bug fix was merged into this PR to avoid conflicts.

## Closes
Closes #2085

### CODE REVIEW

**Comment by IllianiCBT** [comment]
Conflicts resolved

**Comment by IllianiCBT** [comment]
Resolved conflicts

File: MekHQ/src/mekhq/campaign/universe/PlanetarySystem.java [hunk 1]
```diff
@@ -169,28 +170,27 @@ public String getName(LocalDate when) {
...
 
     public Set<Faction> getFactionSet(LocalDate when) {
-        Set<Faction> factions = new HashSet<>();
-        for (Planet planet : planets.values()) {
-            Set<Faction> f = planet.getFactionSet(when);
-            if (null != f) {
-                factions.addAll(f);
-            }
-        }
+        Set<Faction> factions = planets.values().stream()
```
Comment (NickAragua) [inline MekHQ/src/mekhq/campaign/universe/PlanetarySystem.java:183]: I'm kind of curious to see a "how long does this take to run" comparison between this and the older code. In my experience, the syntactic sugar carries a performance penalty, but I'm happy to be proven wrong.

File: MekHQ/src/mekhq/gui/panes/UnitMarketPane.java [hunk 2]
```diff
@@ -469,6 +482,50 @@ private void finalizeEntityAcquisition(final List<UnitMarketOffer> offers,
         getMarketModel().setData(getCampaign().getUnitMarket().getOffers());
     }
 
+    /**
+     * Returns the quality of a unit based on the given market type.
+     *
+     * @param market the type of market
+     * @return the quality of the unit
+     */
+    private int getQuality(UnitMarketType market) {
```
Comment (NickAragua) [inline MekHQ/src/mekhq/gui/panes/UnitMarketPane.java:491]: Do we want to have this "business logic" code in this UI class, or maybe move it to the UnitMarketType class instead?

File: MekHQ/src/mekhq/campaign/market/unitMarket/AtBMonthlyUnitMarket.java [hunk 3]
```diff
@@ -150,8 +239,19 @@ public void addOffers(final Campaign campaign, final int num, UnitMarketType mar
             final Collection<MissionRole> missionRoles = new ArrayList<>();
             if (unitType == UnitType.TANK) {
                 movementModes.addAll(IUnitGenerator.MIXED_TANK_VTOL);
-                missionRoles.add(MissionRole.MIXED_ARTILLERY);
+
+                // should a special unit type be picked? This allows us to force MissionRole that would otherwise be filtered out
+                if (Compute.randomInt(30) == 0) {
```
Comment (NickAragua) [inline MekHQ/src/mekhq/campaign/market/unitMarket/AtBMonthlyUnitMarket.java:244]: How did we arrive at "1 in 30 tanks should be artillery or support"?

**Review by NickAragua — CHANGES_REQUESTED** [review]
A couple of questions/maintainability requests.

File: MekHQ/src/mekhq/campaign/universe/PlanetarySystem.java (hunk 1 above)
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/universe/PlanetarySystem.java:183]: I was under the impression that streams were faster than loops, but that appears to not be as universally true as I was led to believe.

According to google, streams are faster when looping through large volumes of data, but slower with small volumes. This is definitely a case of 'small volume' so I'll go ahead and revert this change and be more judicious with their usage in future.

File: MekHQ/src/mekhq/campaign/market/unitMarket/AtBMonthlyUnitMarket.java (hunk 3 above)
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/market/unitMarket/AtBMonthlyUnitMarket.java:244]: That was just the value that 'felt right' during testing.

There is an argument to be made for making this a campaign option, but given how roles now work it's not possible to use the old method of generation, so I had to pick a value.

After Stucco's role overhaul, the `MIXED_ARTILLERY` MissionRole now filters out anything that isn't mixed artillery (which just happens filter out all vehicles). Previously this role was unique in that it deliberately didn't filter anything out, it just told mhq to include mixed artillery in addition to other units.

Not including some kind of arbitrary system to force the unit market to add artillery (and support) units results in those units never being added to the unit market.  Potentially this is due to filtering occurring elsewhere in the code, but I couldn't find anything that would indicate this taking place.

Comment (NickAragua) [reply MekHQ/src/mekhq/campaign/market/unitMarket/AtBMonthlyUnitMarket.java:244]: Fair enough, we can always revisit it in a subsequent PR.

**Comment by IllianiCBT** [comment]
Updated to include requested revisions.

I also moved the arbitrary 'special vehicle chance' value to campaign options. It's no less arbitrary, but this at least means the user can better adjust it to meet their needs.

Moving `getQuality` was handled in a second follow-up commit, as I forgot to include it in the initial revision. Apologies for the spam.

**Pull request merged**